ambry-router/src/main/java/com.github.ambry.router/GetBlobOperation.java

```diff
@@ -530,12 +541,45 @@ RouterException getChunkException() {
     void poll(RequestRegistrationCallback<GetOperation> requestRegistrationCallback) {
       //First, check if any of the existing requests have timed out.
       cleanupExpiredInFlightRequests();
-      checkAndMaybeComplete();
-      if (!isComplete()) {
-        fetchRequests(requestRegistrationCallback);
+      if (!maybeProcessCallbacksAndCompleteOperation()) {
+        checkAndMaybeComplete();
+        if (!isComplete()) {
+          fetchRequests(requestRegistrationCallback);
+        }
       }
     }
 
+    /**
+     * Maybe process callbacks and complete the operation if applicable. This is a no-op for blobs that doesn't need
+     * any async processing. As of now, decryption is the only async processing that could happen if applicable.
+     * @return {@code true} if the operation is complete, {@code false} otherwise
+     */
+    protected boolean maybeProcessCallbacksAndCompleteOperation() {
+      if (progressTracker.isDecryptionRequired() && decryptCallbackResultInfo.decryptJobComplete) {
+        logger.trace("Processing decrypt callback stored result for data chunk {}", chunkBlobId);
+        if (decryptCallbackResultInfo.exception == null && (getChunkException() == null
+            || !retainChunkExceptionOnSuccess)) {
+          chunkIndexToBuffer.put(chunkIndex,
+              filterChunkToRange(decryptCallbackResultInfo.result.getDecryptedBlobContent()));
+          numChunksRetrieved++;
+          logger.trace("Decrypt result successfully updated for data chunk {}", chunkBlobId);
+          progressTracker.setDecryptionSuccess();
+        } else {
+          logger.trace("Setting operation exception as decryption callback invoked with exception {} for data chunk {}",
+              decryptCallbackResultInfo.exception, chunkBlobId);
+          setOperationException(
+              new RouterException("Exception thrown on decrypting the content for data chunk " + chunkBlobId,
+                  decryptCallbackResultInfo.exception, RouterErrorCode.UnexpectedInternalError));
+          progressTracker.setDecryptionFailed();
+        }
+        logger.trace("Marking blob content available to process for data chunk {}", chunkBlobId);
+        checkAndMaybeComplete();
+        onChunkOperationComplete(this);
```

Comment:
I know you have looked to refactor this a lot, but I kinda feel that you end up calling these here because of the NOT check in the `poll()` which at least now seems unnecessary. Let us sync up today.

Reply:
nope. bcoz of OnChunkOperationComplete(). Will sync up with you offline